### ISSUE TITLE
[Feature] Add global setting to make getByText default to exact: true

### ISSUE THREAD

**Issue by jpveooys** [microsoft/playwright#19235]
[The default behaviour in Testing Library is `{ exact: true }`](https://testing-library.com/docs/queries/bytext), while Playwright defaults to `{ exact: false }`.

Naturally this is bound to catch people out. It would be useful to have at least a global setting to change the default to `{ exact: true }`.

See also https://github.com/microsoft/playwright/discussions/18239

**Comment by Nav-2d** [comment]
Related comment/issue: https://github.com/microsoft/playwright/issues/19046#issuecomment-1327913236

**Comment by mxschmitt** [comment]
> We made `exact: false` the default to align with all other locators. Overall, it is usually more convenient to write lax locators, especially when you do that manually.

Quoting from above as a reply.

**Comment by jpveooys** [comment]
FWIW, our preference is `exact: true` because:

- a case-insensitive substring search can cause unwanted matches e.g. 'Phone number' would match both 'Phone number' and 'Alternative phone number'
- we want our tests to fail if the case of the text in the test doesn't match what's on the page

So, to be honest, the case-insensitivity is probably more the problem than the fact that `{ exact: false }`  does a substring search.

In any case, I did deliberately phrase the issue as a request for a global setting to opt-in to a less lax default, as changing the default would be a breaking change in any case.

**Comment by mxschmitt** [comment]
You can do `/Phone number/` to make it case sensitive. Would that work for you?

**Comment by jpveooys** [comment]
Actually no: I just checked and another case where it tripped up was with `getByLabel` and a checkbox labelled 'Email' and text input named 'Email address' on the same page. So with a regex we'd probably end up doing `/^Phone number$/` anyway.

**Comment by aslushnikov** [comment]
> So with a regex we'd probably end up doing /^Phone number$/ anyway.

@jpveooys this still looks reasonably good to me. Do you find the ergonomics of these regexes sub-par?

**Comment by jpveooys** [comment]
I do, yes. While in most cases the regex would be simple enough, there's still more room for something to go wrong with a regex, and you have to worry about escaping etc. If in one place `/^Phone number$/` was used, and in another `/^Phone number/` was used, is it intentional or a mistake? It's hard to know.

My personal preference is to limit regexes to where they are really needed, so they can get more attention and it's obvious there's something a bit different there.

**Comment by joshsalverda** [comment]
Would also be interested in this. Currently migrating over from Cypress/Cypress Testing Library and really enjoying working with Playwright.

I'd like to be able to select an input by label "Street address" without it also matching "Street address 2", for example. Not a fan of creating the extra regex matchers when just using "exact" would handle all of my use cases already.

In the meantime I'm doing this hacky workaround:

```
baseTest.extend<MyFixtures & MyOptions>({
  page: async ({page}, use) => {
    // Hacky hack to default page locators to use exact matches on strings
    // TODO remove this if/when playwright supports defaulting the "exact" option
    await use(new Proxy(page, {
      get: function (target, prop) {
        if(typeof prop === "string" && prop.startsWith("getBy")) {
          return new Proxy(Reflect.get(target, prop), {
            apply: (target, thisArg, argumentsList) => {
              return Reflect.apply(target, thisArg, [argumentsList[0], { exact: true, ...argumentsList[1] }]);
            }
          });
        }
        return Reflect.get(target, prop);
      }
    }));
  },
  ...other fixtures
});
```

Not 100% sure this works for all cases, but it's working alright for me so far.

**Comment by aslushnikov** [comment]
Everybody: would a `page.getByExactText()` method satisfy your needs here? 

Related: https://github.com/microsoft/playwright/issues/22572

**Comment by joshsalverda** [comment]
@aslushnikov I don't think so, since exact matches also affect getByLabel (and I'm assuming other text matchers). So you're opening the can of worms to add getByExactLabel, getByExactPlaceholder, getByExactTitle, etc.

Although, I mean if that's the API you want to go with then that's fine, just seems like more hassle.

**Comment by jpveooys** [comment]
I agree with @joshsalverda. Although the title of this issue is about `getByText`, `getByLabel` has also been discussed and it's equally valid for the others.

**Comment by klhex** [comment]
> I agree with @joshsalverda. Although the title of this issue is about `getByText`, `getByLabel` has also been discussed and it's equally valid for the others.

I'd prefer @aslushnikov suggestion (or rather the one initiated by #22572) over a global option because this way you would see directly in the code what is intended and you don't have locators behaving different depending on config options you might not be aware of.

Also, the number of getBy-Locators is not that high, I would like having an "exact"-Version of them. Why not start with getByExactText and then see which other ones are most requested thereafter...

**Comment by mrmckeb** [comment]
We've hit this recently too. Whilst we don't disagree with the defaults @mxschmitt, our team are working with both Testing Library (across multiple codebases) and Playwright and this causes additional mental overhead.

Being able to make these work the same way will help engineers be more productive (hitting less errors when writing tests).

Perhaps the global setting here could instead be a `testingLibraryCompat` boolean, in case other inconsistencies appear over time?

**Comment by frontendphil** [comment]
I also want to chime in here and vote **at least** for a global setting. A case-insensitive substring match just led to a test failure when nothing was broken. In my book, this is a strong argument for the whole thing not being the optimal solution. I've introduced a new button, "Save as copy," next to "Save," and now two things match instead of one. I assume that there is some reason that `getByRole('button', { name: "Save" })` would match "Save as copy," but defending that this was indeed a good idea is hard for me.

Especially when this API was (I guess) intended to mimic the one of testing-library. Anyway, if I'd be able to toggle this behavior globally, then I wouldn't have a problem.

**Comment by ghost** [comment]
I totally agree, @tsgautier - please reconsider giving this issue some consideration, seeing that it was opened over three years ago is such a drag ..

**Comment by mrmckeb** [comment]
If we are open to one of the solutions here, I'm happy to contribute that change @mxschmitt @aslushnikov.

I think from what I'm reading above, most people would be happy with a global config to swap the default behaviour of `exact` - aligning with other tools. That seems fairly straight forward, but I may be missing context.

**Comment by ghost** [comment]
This is exactly how I perceived the discussion, and would be the shortest way to a viable, backwards-compatible solution.
